### PR TITLE
feat(noir_js): allow providing foreign call handlers in noirJS

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

This PR allows users to provide an implementation of a foreign call handler when executing circuits using noirJS. The motivation for this is @signorecello getting errors being thrown when executing a circuit which contains `std::println`.

I've neglected providing a default implementation for println as we'd need to replicate the string formatting logic which exists in the Rust code. For now we can just direct users to provide their own foreign call handler and make them responsible for printing results as they like.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/noir_js/src/witness_generation.ts [hunk 1]
```diff
@@ -1,19 +1,21 @@
 import { abiEncode, InputMap } from '@noir-lang/noirc_abi';
 import { base64Decode } from './base64_decode.js';
-import { executeCircuit, WitnessMap } from '@noir-lang/acvm_js';
+import { executeCircuit, WitnessMap, ForeignCallHandler, ForeignCallInput } from '@noir-lang/acvm_js';
 import { CompiledCircuit } from '@noir-lang/types';
 
+const defaultForeignCallHandler: ForeignCallHandler = (name: string, args: ForeignCallInput[]) => {
```
Comment (signorecello) [inline tooling/noir_js/src/witness_generation.ts:6]: Is name the name of the foreign function? It so, maybe later on we can put together some handlers we can match with this

**Comment by signorecello** [comment]
You think it's worth documenting this? If this PR can wait some hours, I can add some note for people to know how to handle it

File: tooling/noir_js/src/witness_generation.ts (hunk 1 above)
Comment (TomAFrench) [reply tooling/noir_js/src/witness_generation.ts:6]: Yep although I'm not wanting to do this in this PR

> I've neglected providing a default implementation for println as we'd need to replicate the string formatting logic which exists in the Rust code. For now we can just direct users to provide their own foreign call handler and make them responsible for printing results as they like.

**Comment by kevaundray** [comment]
@signorecello passing by, can I get an update on this?

**Comment by signorecello** [comment]
Sorry didn't start yet... will do that asap. I'm a bit caught up with EthLondon stuff 😊

**Comment by signorecello** [comment]
I believe anyone can make this doc change though, it's just a function signature + something saying "you can pass a function as the foreign call handler" with an example

**Comment by signorecello** [comment]
@TomAFrench just added the docs, can can you check please?

File: docs/docs/noir_js/reference/01_noirjs.md [hunk 2]
```diff
@@ -81,6 +84,7 @@ async execute(inputs)
 
 ```js
 const { witness, returnValue } = await noir.execute(inputs)
+const { witness, returnValue } = await noir.execute(inputs, (event) => console.log(event))
```
Comment (TomAFrench) [inline docs/docs/noir_js/reference/01_noirjs.md:87]:
```suggestion
const { witness, returnValue } = await noir.execute(inputs, (name, args) => console.log(`Received foreign call ${name} with arguments ${args}`))
```

We should use the correct function signature here. I don't think that example would compile.

File: docs/docs/noir_js/reference/01_noirjs.md [hunk 3]
```diff
@@ -58,17 +58,20 @@ await noirInstance.init();
 
 This async method allows to execute a circuit to get its witness and return value. [`generateFinalProof`](#generatefinalproof) calls it for you, but you can call it directly (i.e. to feed directly to a backend, or to get the return value).
 
+You can optionally provide a foreignCallHandler, to handle functions that should run outside of the prover (i.e. println)
```
Comment (TomAFrench) [inline docs/docs/noir_js/reference/01_noirjs.md:61]:
```suggestion
You can optionally provide a foreignCallHandler, to handle functions that should run outside of the prover (e.g. `std::println`)
```

**Review by signorecello — APPROVED** [review]

File: tooling/noir_js/src/witness_generation.ts (hunk 1 above)
Comment (kevaundray) [reply tooling/noir_js/src/witness_generation.ts:6]: can we open an issue for this @TomAFrench ?

**Comment by kevaundray** [comment]
Merging please open an issue for https://github.com/noir-lang/noir/pull/3294#discussion_r1376858309

**Pull request merged**